cpp/src/datetime/datetime_ops.cu

```diff
@@ -0,0 +1,175 @@
+
+  typedef enum {
+    invalid = 0,
+    year,
+    month,
+    day,
+    weekday,
+    hour,
+    minute,
+    second,
+  } datetime_component;
+
+  template <datetime_component DatetimeComponent>
+  struct extract_component_operator {
+
+    column_device_view column;
+
+    CUDA_HOST_DEVICE_CALLABLE extract_component_operator(column_device_view col) : column(col) {}
+
+    template <typename Element>
+    typename std::enable_if_t<not cudf::is_timestamp_t<Element>::value, int16_t>
+    CUDA_HOST_DEVICE_CALLABLE operator()(size_type const i) const { return 0; }
+
+    template <typename Element>
+    typename std::enable_if_t<cudf::is_timestamp_t<Element>::value, int16_t>
+    CUDA_HOST_DEVICE_CALLABLE operator()(size_type const i) const {
+
+      using namespace simt::std::chrono;
+      using Duration = typename Element::duration;
+
+      auto ts = column.element<Element>(i);
+      auto days_since_epoch = floor<days>(ts);
+
+      switch (DatetimeComponent) {
+        case cudf::datetime::detail::year: return static_cast<int>(year_month_day(days_since_epoch).year());
+        case cudf::datetime::detail::month: return static_cast<unsigned>(year_month_day(days_since_epoch).month());
+        case cudf::datetime::detail::day: return static_cast<unsigned>(year_month_day(days_since_epoch).day());
+        case cudf::datetime::detail::weekday: return year_month_weekday(days_since_epoch).weekday().iso_encoding();
+        default: break;
+      }
+
+      auto time_since_midnight = ts - days_since_epoch;
+
+      if (time_since_midnight.count() < 0) {
+        time_since_midnight += days(1);
+      }
+
+      auto hrs_ = duration_cast<hours>(time_since_midnight);
+      auto mins_ = duration_cast<minutes>(time_since_midnight - hrs_);
+      auto secs_ = duration_cast<seconds>(time_since_midnight - hrs_ - mins_);
+
+      switch (DatetimeComponent) {
+        case cudf::datetime::detail::hour: return hrs_.count();
+        case cudf::datetime::detail::minute: return mins_.count();
+        case cudf::datetime::detail::second: return secs_.count();
+        default: return 0;
+      }
+    }
+  };
+
+  template <datetime_component DatetimeComponent>
+  struct launch_extract_component {
+
+    column_device_view input;
+    mutable_column_view output;
+
+    launch_extract_component(column_device_view inp, mutable_column_view out) : input(inp), output(out) {}
+
+    template <typename Element>
+    typename std::enable_if_t<not cudf::is_timestamp_t<Element>::value, void>
+    operator()(cudaStream_t stream) {
+      CUDF_FAIL("Cannot extract datetime component from non-timestamp column.");
+    }
+
+    template <typename Element>
+    typename std::enable_if_t<cudf::is_timestamp_t<Element>::value, void>
+    operator()(cudaStream_t stream) {
+      auto functor = extract_component_operator<DatetimeComponent>{input};
+      auto bound_f = [=] __device__ (size_type const i) {
+        return functor.template operator()<Element>(i);
```

Comment:
Instead of making `Element` a template of `extract_component_operator::operator()`, how about make it a template parameter of `extract_component_operator`?

So instead this would look like:

```
thrust::tabulate(..., output.begin<int16_t>, output.end<int16_t>, extract_component_operator<DatetimeComponent, Element>{input});
```